netuitive/element.py

```diff
@@ -134,23 +143,15 @@ def add_sample(self,
 
         metricIdSan = self._sanitize(metricId)
 
-        if len(self.metrics) > 0:
-            t = list(self.metrics)
+        if self._metrics is None:
+            self._metrics = {}
 
-            if metricIdSan not in t:
-                self.metrics.append(
-                    Metric(metricIdSan,
-                           metricType,
-                           sparseDataStrategy,
-                           unit,
-                           Tags))
-        else:
-            self.metrics.append(
-                Metric(metricIdSan,
-                       metricType,
-                       sparseDataStrategy,
-                       unit,
-                       Tags))
+        if self._metrics.get(metricIdSan) is None:
```

Comment:
This is the core of the improvement, correct? This get operation is much faster than a lookup in an array?

Reply:
I already explained that in the commit message, it's O(1) v.s. O(n)

Reply:
I also think the original "metricId not in list(metrics)" is a bug, since metricId is a string, list(metrics) is a list of Metric objects, that if will always returns true.

Reply:
Wonderful, looking forward to seeing these improvements.